### PR TITLE
04. Crypto - Scheduler task to continue migration in case of server crash

### PR DESCRIPTION
**Pull request by stefania-santimbrean**
Add in config file:
```
	"Scheduler": {
		"active": true,
		"tasks": [
			{...},
			{
				"name": "MigrateSensitiveDataTask",
				"active": true,
				"periodicity": "* * * * *",
				"config": {}
			}
		]
	}
```

### CODE REVIEW

File: src/utils/Cypher.ts [hunk 1]
```diff
@@ -145,4 +139,111 @@ export default class Cypher {
...
+    // Filter settings with sensitiveData
+    return settings.result.filter((value: SettingDB) => {
+      if (value?.sensitiveData && !Utils.isEmptyArray(value?.sensitiveData)) {
+        return true;
+      }
+    });
+  }
+
+  public static async migrateSettings(tenantID: string, cryptoSetting: CryptoSetting): Promise<void> {
+
```
Comment (jerome-benoit) [inline src/utils/Cypher.ts:185]: You can remove blank line.

File: src/utils/Cypher.ts [hunk 2]
```diff
@@ -145,4 +139,111 @@ export default class Cypher {
...
+    const settingsToMigrate = await Cypher.getSettingsWithSensitiveData(tenantID);
+    // If tenant has settings with sensitive data, migrate them
+    if (!Utils.isEmptyArray(settingsToMigrate)) {
+      // Migrate
+      for (const setting of settingsToMigrate) {
+        if (!setting.formerSensitiveData && Utils.isEmptyArray(setting.formerSensitiveData)) {
+          delete setting.formerSensitiveData;
+          // Save former senitive data in setting
```
Comment (jerome-benoit) [inline src/utils/Cypher.ts:193]: typo: sensitive

File: src/utils/Cypher.ts [hunk 3]
```diff
@@ -1,44 +1,38 @@
...
-          method: 'getConfiguration',
-          message: 'Crypto configuration is missing'
-        });
-      }
-    }
-    return this.configuration;
-  }
 
-  public static encrypt(data: string): string {
+  public static async encrypt(data: string, tenantID: string, former = false): Promise<string> {
```
Comment (jerome-benoit) [inline src/utils/Cypher.ts:17]: I do not really like the signature change here. A better signature would have been to just pass the key to use as an argument.

File: src/utils/Cypher.ts [hunk 4]
```diff
@@ -145,4 +139,111 @@ export default class Cypher {
...
+  }
+
+  public static async migrateSettings(tenantID: string, cryptoSetting: CryptoSetting): Promise<void> {
+
+    const settingsToMigrate = await Cypher.getSettingsWithSensitiveData(tenantID);
+    // If tenant has settings with sensitive data, migrate them
+    if (!Utils.isEmptyArray(settingsToMigrate)) {
+      // Migrate
+      for (const setting of settingsToMigrate) {
+        if (!setting.formerSensitiveData && Utils.isEmptyArray(setting.formerSensitiveData)) {
```
Comment (jerome-benoit) [inline src/utils/Cypher.ts:191]: I do not understand that test: if there's no formerSensitiveData (or an empty array of them), why launching the migration?

The sanity check is that at key change, if there's a previous former settings, they shall be replaced: current -> former, previous former -> deleted and new key is added.

**Review by jerome-benoit — CHANGES_REQUESTED** [review]

File: src/utils/Cypher.ts (hunk 3 above)
Comment (stefania-santimbrean) [reply src/utils/Cypher.ts:17]: Yes, I agree. But in a previous review you suggested putting the tenantID in the encrypt/decrypt and making a getCryptoKey private method that we only call in Cypher.  But even if I pass the cryptoKey (which contains both key, keyPropeties, formerKey, formerKeyProperties), I still have the same problem. Is it better to create a new type which would contain only key and keyProperties and pass that to the encrypt/decrypt?

File: src/utils/Cypher.ts (hunk 4 above)
Comment (stefania-santimbrean) [reply src/utils/Cypher.ts:191]: After a migration is finished the formerSensitiveData are cleaned up from settings. Which means that if the setting does not have formerSensitiveData (and migrationToBeDone is true), that setting has to be migrated - and the opposite if it has formerSensitiveData (and migrationToBeDone is true), that settings was migrated. After a migration is completed, the formerSensitiveData is cleaned and migrationToBeDone is set to false

File: src/utils/Cypher.ts (hunk 3 above)
Comment (jerome-benoit) [reply src/utils/Cypher.ts:17]: Fair enough. Then just rename the boolean argument to something more meaningful like `useFormerKey`.

File: src/utils/Cypher.ts (hunk 4 above)
Comment (jerome-benoit) [reply src/utils/Cypher.ts:190]: > After a migration is finished the formerSensitiveData are cleaned up from settings. Which means that if the setting does not have formerSensitiveData (and migrationToBeDone is true), that setting has to be migrated - and the opposite if it has formerSensitiveData (and migrationToBeDone is true), that settings was migrated. After a migration is completed, the formerSensitiveData is cleaned and migrationToBeDone is set to false

The migration code shall be reentrant, with reentrant code, you they only need to test for migrationToBeDone and then run again the code. Testing for the migration status is then not needed.

File: src/utils/Cypher.ts (hunk 3 above)
Comment (stefania-santimbrean) [reply src/utils/Cypher.ts:17]: I renamed in PR 03. Crypto

File: src/utils/Cypher.ts (hunk 4 above)
Comment (stefania-santimbrean) [reply src/utils/Cypher.ts:190]: I don't understand what should I change? Testing migrationToBeDone is needed to know if migration is finished or not. Testing for formerSensitiveData is needed to know if a particular setting what migrated or not.

**Review by LucasBrazi06 — APPROVED** [review]

**Comment by LucasBrazi06** [comment]
@jerome-benoit 
Could you please update AWS and SCP config with this new job?
```
{
  name": "MigrateSensitiveDataTask",
  "active": true,
  "periodicity": "* * * * *",
  "config": {}
}
```
Thanks.

**Pull request merged**